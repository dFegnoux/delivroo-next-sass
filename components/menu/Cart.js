import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import Link from "next/link";
import QuantityModifier from "./QuantityModifier";
import {
  calculateCartTotalPrice,
  formatToPriceString
} from "../../helpers/cartHelpers";
import "./cart.scss";

class Cart extends Component {
  static propTypes = {
    menus: PropTypes.object.isRequired,
    updateCart: PropTypes.func,
    disableButtons: PropTypes.bool
  };

  static defaultProps = {
    disableButtons: true,
    updateCart: () => {
      console.warning(
        "Trying to update with the default updateCart function, you may want to use a real function ;)"
      );
    }
  };

  render() {
    const { menus, updateCart, disableButtons } = this.props;
    const selectedMenus = Object.entries(menus);
    const isBtnDisabled = !selectedMenus.length ? "disabled" : null;

    return (
      <div className="cartWrapper">
        <div className="cart roundedItem">
          {!disableButtons && (
            <Link href="/cart/validation">
              <a className="btn" disabled={isBtnDisabled}>
                Valider mon panier
              </a>
            </Link>
          )}
          {selectedMenus.length ? (
            <Fragment>
              <ul>
                {selectedMenus.map(menu => (
                  <li key={menu[0]}>
                    <QuantityModifier
                      item={menu[1]}
                      updateCart={updateCart}
                      disabled={disableButtons}
                    />
                    <span className="itemName">{menu[1].name}</span>
                    <span className="total">
                      {formatToPriceString(menu[1].price * menu[1].quantity)}
                    </span>
                  </li>
                ))}
              </ul>
              <div className="mainTotal">
                <span>Total : </span>
                <span>{calculateCartTotalPrice(menus)}</span>
              </div>
            </Fragment>
          ) : (
            <div>Votre panier est vide</div>
          )}
        </div>
      </div>
    );
  }
}

export default Cart;
