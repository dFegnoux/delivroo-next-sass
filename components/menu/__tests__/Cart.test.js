import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import Cart from "../Cart";

const mockedMenus = {
  menu1: {
    id: "menu1",
    title: "title menu1",
    description: "desc menu1",
    price: "15.00",
    picture: "//www.placekitten.com/100/100",
    quantity: 1
  },
  menu2: {
    id: "menu2",
    title: "title menu2",
    description: "desc menu2",
    price: "5.00",
    picture: "//www.placekitten.com/100/100",
    quantity: 3
  }
};
const mockUpdateCart = jest.fn();
const props = {
  menus: mockedMenus,
  updateCart: mockUpdateCart,
  disableButtons: false
};
const container = shallow(<Cart {...props} />);

it("renders correctly", () => {
  const tree = renderer.create(<Cart {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("should have as many list item as menus", () => {
  expect(container.find("li").length).toBe(2);
});

it("should have an updateCart default function", () => {
  expect(Cart.defaultProps.updateCart).toBeDefined();
});

describe("Validate button", () => {
  afterEach(() => {
    container.setProps(props);
  });

  it("should be displayed if props say so", () => {
    expect(container.find("Button").length).toBe(1);
    container.setProps({ disableButtons: true });
    expect(container.find("Button").length).toBe(0);
  });

  it("should be displayed as disabled with an empty cart", () => {
    container.setProps({ menus: {} });
    expect(container.find("Button").props().disabled).toBeTruthy();
  });
});

describe("When Cart is instanciate without UpdateCart function", () => {
  const noFnContainer = shallow(<Cart menus={{}} />);

  it("should render correctly", () => {
    const snapshot = renderer.create(<Cart menus={{}} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it("should warn user when trying to use default update function", () => {
    noFnContainer.instance().props.updateCart();
    expect(console.warning).toHaveBeenCalled();
  });
});
