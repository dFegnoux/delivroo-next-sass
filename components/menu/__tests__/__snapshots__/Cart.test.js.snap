// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Cart is instanciate without UpdateCart function should render correctly 1`] = `
<div
  className="cartWrapper"
>
  <div
    className="cart roundedItem"
  >
    <div>
      Votre panier est vide
    </div>
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="cartWrapper"
>
  <div
    className="cart roundedItem"
  >
    <button
      className="btn [object Object] fullWidth"
      disabled={false}
      onClick={[Function]}
    >
      Valider mon panier
    </button>
    <ul>
      <li>
        <div
          className="quantityModifier"
        >
          <button
            className="btn [object Object] remove small"
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="quantity"
          >
            1
          </span>
          <button
            className="btn [object Object] add small"
            onClick={[Function]}
          >
            +
          </button>
        </div>
        <span
          className="itemName"
        />
        <span
          className="total"
        >
          15,00 €
        </span>
      </li>
      <li>
        <div
          className="quantityModifier"
        >
          <button
            className="btn [object Object] remove small"
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="quantity"
          >
            3
          </span>
          <button
            className="btn [object Object] add small"
            onClick={[Function]}
          >
            +
          </button>
        </div>
        <span
          className="itemName"
        />
        <span
          className="total"
        >
          15,00 €
        </span>
      </li>
    </ul>
    <div
      className="mainTotal"
    >
      <span>
        Total : 
      </span>
      <span>
        30,00 €
      </span>
    </div>
  </div>
</div>
`;
